### PR TITLE
Console add stripe service

### PR DESCRIPTION
**Pull request by rikysya**
What: add stripe service to the console

Why: first part of the https://github.com/storj/storj/pull/2038

Thanks for submitting a PR!

## Code Review Checklist (to be filled out by reviewer)
 - [ ] Does the PR describe what changes are being made?
 - [ ] Does the PR describe why the changes are being made?
 - [ ] Does the code follow [our style guide](https://github.com/storj/docs/blob/master/code/Style.md)?
 - [ ] Does the code follow [our testing guide](https://github.com/storj/docs/blob/master/code/Testing.md)?
 - [ ] Could the PR be broken into smaller PRs?
 - [ ] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [ ] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [ ] Does any documentation need updating?

### CODE REVIEW

File: internal/payments/service.go [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+// GetCustomer retrieves customer object from stripe network
+func (s *StripeService) GetCustomer(ctx context.Context, id string) (*stripe.Customer, error) {
+	return s.client.Customers.Get(id, nil)
+}
+
+// GetCustomerDefaultPaymentMethod retrieves customer default payment method from stripe network
+func (s *StripeService) GetCustomerDefaultPaymentMethod(ctx context.Context, customerID string) (*stripe.PaymentMethod, error) {
+	cus, err := s.client.Customers.Get(customerID, nil)
+	if err != nil {
+		return nil, err
```
Comment (VinozzZ) [inline internal/payments/service.go]: Should we have a `paymentErr` to wrap this?

File: internal/payments/service.go [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+	for iterator.Next() {
+		pm := iterator.PaymentMethod()
+		paymentMethods = append(paymentMethods, pm)
+	}
+
+	return paymentMethods, nil
+}
+
+// GetPaymentMethod retrieve payment method object from stripe network
+func (s *StripeService) GetPaymentMethod(ctx context.Context, id string) (*stripe.PaymentMethod, error) {
```
Comment (VinozzZ) [inline internal/payments/service.go]: Is the `id` here a paymentMethod id from stripe?

File: internal/payments/service.go [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+// - Egress
+// - ObjectsCount
+// Created invoice has AutoAdvance property set to true, so it will be finalized
+// (no further editing) and attempted to be paid in 1 hour after creation
+func (s *StripeService) CreateProjectInvoice(ctx context.Context, params CreateProjectInvoiceParams) (*stripe.Invoice, error) {
+	// create line items
+	_, err := s.client.InvoiceItems.New(&stripe.InvoiceItemParams{
+		Customer:    stripe.String(params.CustomerID),
+		Description: stripe.String("Storage"),
+		Quantity:    stripe.Int64(int64(params.Storage)),
```
Comment (VinozzZ) [inline internal/payments/service.go]: Will there ever be a chance that `params.Storage` overflow in the conversion here?

File: internal/payments/service.go [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+		Currency:    stripe.String(string(stripe.CurrencyUSD)),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = s.client.InvoiceItems.New(&stripe.InvoiceItemParams{
+		Customer:    stripe.String(params.CustomerID),
+		Description: stripe.String("Egress"),
+		Quantity:    stripe.Int64(int64(params.Egress)),
```
Comment (VinozzZ) [inline internal/payments/service.go]: same as above

File: satellite/console/service.go [hunk 5]
```diff
@@ -55,37 +56,36 @@ const (
...
-
 	if store == nil {
 		return nil, errs.New("store can't be nil")
 	}
-
 	if log == nil {
 		return nil, errs.New("log can't be nil")
 	}
-
 	if passwordCost == 0 {
```
Comment (VinozzZ) [inline satellite/console/service.go]: I see that we are checking all the arguments against its zero value here. Do we need to also add the check for `stripeService`?

File: satellite/console/service.go [hunk 6]
```diff
@@ -119,22 +119,41 @@ func (s *Service) CreateUser(ctx context.Context, user CreateUser, tokenSecret R
...
-		Email:        user.Email,
-		FullName:     user.FullName,
-		ShortName:    user.ShortName,
-		PasswordHash: hash,
-	})
+	// store data
+	tx, err := s.store.BeginTx(ctx)
 	if err != nil {
-		return nil, errs.New(internalErrMsg)
+		return nil, err
```
Comment (VinozzZ) [inline satellite/console/service.go]: Is there a reason why we aren't wrapping this error?

File: internal/payments/service.go (hunk 2 above)
Comment (rikysya) [reply internal/payments/service.go]: yes

File: satellite/console/service.go (hunk 5 above)
Comment (rikysya) [reply satellite/console/service.go]: I don't think we need it for internal pkg

File: internal/payments/service.go (hunk 1 above)
Comment (rikysya) [reply internal/payments/service.go]: I am not sure

File: internal/payments/service.go (hunk 3 above)
Comment (rikysya) [reply internal/payments/service.go]: I don't think anyone can store more than `9,223,372,036,854,775,807` `GB` in the nearest future

File: satellite/console/service.go (hunk 6 above)
Comment (rikysya) [reply satellite/console/service.go]: that's not wrapping, but replacing error contents completely with user friendly, but non-descriptive error. I am not fan of this approach, but if needed it can be replaced in the future after some testing period, because it's much easier to identify errors/bugs with real error msgs

File: internal/payments/logger.go [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+import (
+	"fmt"
+
+	"github.com/stripe/stripe-go"
+	"go.uber.org/zap"
+)
+
+// zapWrapper is a wrapper for *zap.Logger.
+// Implements stripe.LeveledLoggerInterface
+type zapWrapper struct {
```
Comment (stefanbenten) [inline internal/payments/logger.go]: Couldnt we get away with a Sugared Logger?
See below implementation proposal

File: internal/payments/service.go [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+	Egress      float64
+	ObjectCount float64
+
+	StartDate time.Time
+	EndDate   time.Time
+}
+
+// NewService creates new instance of StripeService initialized with API key
+func NewService(log *zap.Logger, apiKey string) *StripeService {
+	stripe.DefaultLeveledLogger = wrapLogger(log)
```
Comment (stefanbenten) [inline internal/payments/service.go]: stripe.DefaultLeveledLogger = log.Sugar().Named("") should be valid (Didnt test it tho).
It contains the necessary interface functions.

File: internal/payments/service.go [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+// CreateCustomer creates new customer from CustomerParams struct
+// sets default payment to one of the predefined testing VISA credit cards
+func (s *StripeService) CreateCustomer(ctx context.Context, params CreateCustomerParams) (*stripe.Customer, error) {
+	cparams := &stripe.CustomerParams{
+		Email:       stripe.String(params.Email),
+		Name:        stripe.String(params.Name),
+		Description: stripe.String(params.Description),
+	}
+
+	// Set default source (payment instrument)
```
Comment (stefanbenten) [inline internal/payments/service.go]: Remove commented code

File: internal/payments/service.go (hunk 1 above)
Comment (stefanbenten) [reply internal/payments/service.go]: I like that idea, the more clear the log messages, the better!

File: satellite/console/service.go [hunk 10]
```diff
@@ -55,37 +56,36 @@ const (
...
+	Signer
+	store         DB
+	stripeService payments.Service
 
 	passwordCost int
 }
 
 // NewService returns new instance of Service
-func NewService(log *zap.Logger, signer Signer, store DB, passwordCost int) (*Service, error) {
+func NewService(log *zap.Logger, signer Signer, store DB, stripeService payments.Service, passwordCost int) (*Service, error) {
```
Comment (stefanbenten) [inline satellite/console/service.go]: What happens if a particular Satellite Operator dont want to or can not use Stripe?
Can the satellite be operated without it?

File: satellite/console/service.go [hunk 11]
```diff
@@ -55,37 +56,36 @@ const (
 
 // Service is handling accounts related logic
 type Service struct {
-	Signer
+	log *zap.Logger
 
-	store DB
-	log   *zap.Logger
+	Signer
```
Comment (stefanbenten) [inline satellite/console/service.go]: Can we split external from internal vars into blocks?
Dont think its necessary, but makes the code clearer imo

File: satellite/console/service.go (hunk 10 above)
Comment (rikysya) [reply satellite/console/service.go]: Honestly, I don't know. When we discussed payments we wanted to use stripe wherever we can, so decoupling it gonna break everything

File: satellite/console/service.go (hunk 6 above)
Comment (VinozzZ) [reply satellite/console/service.go]: That makes sense. How about using the `Error = errs.Class("console error")` and then `Error.Wrap(err)`. I saw this pattern in other services

File: satellite/console/service.go [hunk 12]
```diff
@@ -119,22 +119,41 @@ func (s *Service) CreateUser(ctx context.Context, user CreateUser, tokenSecret R
...
+		return nil, err
 	}
 
-	err = s.store.RegistrationTokens().UpdateOwner(ctx, registrationToken.Secret, u.ID)
-	if err != nil {
-		return nil, errs.New(internalErrMsg)
-	}
+	// clean up user on error
+	defer func() {
+		if err != nil {
```
Comment (jtolds) [inline satellite/console/service.go]: this is pretty scary. this will suddenly start having bugs if for some reason someone starts changing the names of the return values. can you put this defer at the very top of the function before any errors are defined so it's clear we mean for it to refer to the named return value?

File: satellite/console/consoleweb/consoleql/query_test.go [hunk 13]
```diff
@@ -20,6 +20,7 @@ import (
 	"storj.io/storj/satellite/console/consoleauth"
 	"storj.io/storj/satellite/console/consoleweb/consoleql"
 	"storj.io/storj/satellite/mailservice"
+	"storj.io/storj/satellite/payments/internalservice"
```
Comment (jtolds) [inline satellite/console/consoleweb/consoleql/query_test.go]: can we name this package something more descriptive? `internalservice` is surprising

File: satellite/console/service.go [hunk 14]
```diff
@@ -119,22 +119,41 @@ func (s *Service) CreateUser(ctx context.Context, user CreateUser, tokenSecret R
...
+			return errs.New(internalErrMsg)
+		}
+
+		err = tx.RegistrationTokens().UpdateOwner(ctx, registrationToken.Secret, u.ID)
+		if err != nil {
+			return errs.New(internalErrMsg)
+		}
+
+		return nil
+	})
```
Comment (jtolds) [inline satellite/console/service.go]: why is this create user stuff changing in this PR at all? isn't this PR about stripe?

one thing you can do to break up commits is use `git add -p` which is very helpful. it allows you to add only some changes in a file. is this change stripe related?

File: satellite/console/service.go [hunk 15]
```diff
@@ -441,39 +460,38 @@ func (s *Service) CreateProject(ctx context.Context, projectInfo ProjectInfo) (p
...
-	}
-
-	transaction, err := s.store.BeginTx(ctx)
+	tx, err := s.store.BeginTx(ctx)
 	if err != nil {
 		return nil, errs.New(internalErrMsg)
 	}
 
+	// clean up project on error
 	defer func() {
```
Comment (jtolds) [inline satellite/console/service.go]: same about this defer. if someone changes the named return value names, everything will still compile, but this defer method will stop working right. instead, if this was at the top of the function, then it wouldn't compile if someone changed the names.

File: satellite/payments/internalservice/internal.go [hunk 16]
```diff
@@ -0,0 +1,110 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package internalservice
```
Comment (jtolds) [inline satellite/payments/internalservice/internal.go]: maybe this is called "internalpayments" or "localpayments"?

File: satellite/console/service.go (hunk 14 above)
Comment (rikysya) [reply satellite/console/service.go]: as this is only part of original PR, those changes were added later with more stripe related stuff. And when picking changes to commit I accidently picked them up to this PR
